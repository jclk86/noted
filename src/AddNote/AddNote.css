label,
input {
  display: block;
}

textarea {
  display: block;
  margin: 0 auto;
  width: 400px;
  height: 200px;
}

.button-container {
  text-align: center;
}

form {
  text-align: left;
}

select {
  display: block;
  margin: 10px auto;
}

.error-message {
  display: inline;
}

.add-content-label {
  margin-top: 40px;
}

.add-content-container {
  text-align: center;
}

input[name="name"] {
  width: 300px;
}
