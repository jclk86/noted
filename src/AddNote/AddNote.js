import React from "react";
import "./AddNote.css";
import Context from "../context";
import Config from "../config";
import ValidationError from "../ValidationError";
import { withRouter } from "react-router-dom";

class AddNote extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      name: {
        value: "",
        touched: false
      },
      content: {
        value: "",
        filled: false
      },
      folderId: {
        value: "",
        chosen: false
      },
      modified: new Date()
    };
  }
  static contextType = Context;

  validateName = fieldValue => {
    const name = this.state.name.value.trim();
    if (name.length === 0) {
      return "Please enter a name for the note";
    } else if (name.length < 3 || name.length > 20) {
      return "Note name should be between 3 and 20 characters";
    }
  };

  validateContent = fieldValue => {
    const content = this.state.content.value;
    if (content.length === 0) {
      return "Please enter content";
    }
  };

  validateFolderId = fieldValue => {
    const folderId = this.state.folderId.value;
    if (folderId === null) {
      return "Please select a folder";
    }
  };

  getTitle = name => {
    this.setState({
      name: { value: name, touched: true }
    });
  };

  getContent = content => {
    this.setState({
      content: { value: content, filled: true }
    });
  };

  getFolderId = folderId => {
    this.setState({
      folderId: {
        value: folderId,
        chosen: true
      }
    });
  };

  isFormValid = () => {
    const { name, content, folderId } = this.state;
    return name.value && content.value && folderId.value;
  };
  // add body: JSON.stringify
  handleSubmit = event => {
    event.preventDefault();
    const note = {
      name: event.target.name.value,
      modified: this.state.modified,
      folderId: event.target.folderId.value,
      content: event.target.content.value
    };
    fetch(`${Config.API_ENDPOINT}/notes`, {
      method: "post",
      body: JSON.stringify(note),
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json"
      }
    })
      .then(res => {
        if (!res.ok) {
          throw new Error("Oops... Something went wrong");
        }
        return res.json();
      })
      .then(note => {
        this.context.addNote(note);
        this.props.history.goBack();
      })
      .catch(e => {
        console.error(e);
      });
  };

  render() {
    const valid = this.isFormValid();
    return (
      <div className="add-note-container">
        <form onSubmit={e => this.handleSubmit(e)}>
          <div className="add-name-container">
            <label htmlFor="add name" className="add-note-label" /> Add Title:
            <input
              type="text"
              id="name"
              name="name"
              placeholder="My To Dos"
              onChange={e => this.getTitle(e.target.value)}
            />
          </div>
          {this.state.name.touched && (
            <ValidationError message={this.validateName()} />
          )}
          <div className="add-content-container">
            <label className="add-content-label" htmlFor="add content" /> Add
            Content:
            <textarea
              id="content"
              name="content"
              placeholder="...buy groceries"
              onChange={e => this.getContent(e.target.value)}
            />
            {!this.state.content.value && (
              <ValidationError message={this.validateContent()} />
            )}
          </div>
          <select
            name="folderId"
            onChange={e => this.getFolderId(e.target.value)}
          >
            <option value={""}>Select Folder</option>
            {this.context.folders.map(folder => (
              <option key={folder.id} value={folder.id}>
                {folder.name}
              </option>
            ))}
          </select>
          {!this.state.folderId.chosen && (
            <ValidationError message={this.validateFolderId()} />
          )}
          <div className="button-container">
            <button type="submit" disabled={!valid}>
              Submit
            </button>
          </div>
          <div className="note-return-button">
            <button type="button" onClick={this.props.history.goBack}>
              Go Back
            </button>
          </div>
        </form>
      </div>
    );
  }
}

export default withRouter(AddNote);
