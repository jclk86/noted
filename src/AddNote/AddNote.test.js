import React from "react";
import { shallow } from "enzyme";
import { toJson } from "enzyme-to-json";
import AddNote from "./AddNote";
import Context from "../context";

describe("AddNote Component", () => {
  const contextType = {
    addNote: () => {},
    folders: [
      {
        id: "d26e0bce-ffaf-11e8-8eb2-f2801f1b9fd1",
        name: "Elephants",
        modified: "2018-04-11T23:00:00.000Z",
        folderId: "b07161a6-ffaf-11e8-8eb2-f2801f1b9fd1",
        content:
          "Officiis exercitationem minus. Officiis et ex debitis iusto. Possimus est et suscipit a nesciunt laboriosam. Accusamus recusandae vel architecto velit illum labore dolor ipsam nulla.\n \rEveniet numquam non et cumque incidunt veniam et. Quibusdam voluptates neque et autem. Omnis ea sunt delectus.\n \rCorrupti et qui sit enim. Saepe sapiente est eveniet ut debitis et nulla. In doloribus dolores dolor voluptas commodi eligendi qui qui."
      },
      {
        id: "d26e0f48-ffaf-11e8-8eb2-f2801f1b9fd1",
        name: "Lions",
        modified: "2018-04-26T23:00:00.000Z",
        folderId: "b0715efe-ffaf-11e8-8eb2-f2801f1b9fd1",
        content:
          "Sunt veritatis vero impedit. Perferendis quis nostrum. Earum est nesciunt.\n \rIllum nemo architecto pariatur pariatur repellendus sint consequatur quia voluptates. Maxime reiciendis voluptatibus amet. Consectetur voluptas minima sit delectus. Quos suscipit et neque. Nihil pariatur dicta aliquid sunt voluptatem. Ea dolores eos eius ipsa ratione iusto qui deleniti sint.\n \rEt repellat sequi eaque doloribus qui ab quam ipsum impedit. Voluptatem ut soluta odio sit ratione dolore. Vero aut ipsa cumque ratione corrupti ratione aut temporibus debitis."
      },
      {
        id: "d26e1074-ffaf-11e8-8eb2-f2801f1b9fd1",
        name: "Monkeys",
        modified: "2018-02-05T00:00:00.000Z",
        folderId: "b07162f0-ffaf-11e8-8eb2-f2801f1b9fd1",
        content:
          "Optio voluptatem dignissimos. Ex qui optio et eum est sunt hic. Soluta saepe voluptas occaecati dolorem quasi.\n \rAb non quas reiciendis voluptates eius error. Accusamus omnis aliquid hic ratione fugiat molestiae voluptatem. Voluptatem sed ipsam. Quae et laboriosam.\n \rPerferendis ad recusandae magnam quo. Iure quis pariatur distinctio vero. Voluptas non aut voluptas quia optio qui aliquam temporibus numquam. Quis odit recusandae et quia."
      },
      {
        id: "d26e11a0-ffaf-11e8-8eb2-f2801f1b9fd1",
        name: "Bats",
        modified: "2018-12-01T00:00:00.000Z",
        folderId: "b0715efe-ffaf-11e8-8eb2-f2801f1b9fd1",
        content:
          "Et vel inventore. Dolores consectetur adipisci eius cum fuga excepturi et voluptatem. Et amet et enim facere facere velit. Voluptatem nemo officiis. Aut eligendi voluptatem voluptas et voluptatem saepe laudantium dolores voluptas. Quis enim quibusdam quis exercitationem necessitatibus sapiente.\n \rRepellendus quibusdam earum autem repellendus quis perferendis id vitae. Dicta laudantium consequuntur et sint id. Sapiente et qui voluptates est. Qui alias ex provident assumenda distinctio nihil.\n \rEarum nihil et eligendi esse quod. Est non ipsum totam aliquid quam sequi voluptate repellendus. Aut animi officiis quo."
      },
      {
        id: "d26e12c2-ffaf-11e8-8eb2-f2801f1b9fd1",
        name: "Turtles",
        modified: "2018-09-11T23:00:00.000Z",
        folderId: "b07161a6-ffaf-11e8-8eb2-f2801f1b9fd1",
        content:
          "Voluptatum velit et culpa omnis voluptas modi aut neque. Aut expedita facere dolores dolores. Qui non voluptatum quia eligendi consequuntur enim. Sit qui sit ex ipsum perspiciatis qui delectus omnis beatae. Quos quia quod sunt.\n \rCorrupti consequatur voluptatem eveniet tenetur cumque eos quia eius est. Deleniti sit repellendus ipsam nihil eveniet. Facere aperiam totam eos praesentium reprehenderit. Est qui quod minus possimus. Facere voluptatum minus tenetur. Quae aut mollitia aut eveniet eius expedita qui.\n \rQuia quam molestiae ut itaque maxime blanditiis. Voluptatum mollitia exercitationem accusamus. Omnis natus mollitia aliquam illum dolorem."
      },
      {
        id: "d26e1452-ffaf-11e8-8eb2-f2801f1b9fd1",
        name: "Zebras",
        modified: "2018-08-13T23:00:00.000Z",
        folderId: "b07161a6-ffaf-11e8-8eb2-f2801f1b9fd1",
        content:
          "Aspernatur repudiandae sunt recusandae animi. Cumque eveniet praesentium. Perspiciatis pariatur ducimus nostrum maxime. Adipisci vel minima. Nostrum corrupti quae.\n \rEst perspiciatis ipsa. Illum expedita omnis magni atque ea. Voluptas consequatur consequuntur est earum nihil nihil. Ratione ut consequatur quo. Possimus eaque quidem ipsum voluptas.\n \rIllum aperiam iste qui dicta et consequatur. Et in necessitatibus odit vitae at necessitatibus eveniet ducimus facilis. Ad ut eius officiis ut dignissimos dolorum. Velit velit et. Vero incidunt adipisci quis voluptates expedita quo. Tempore eum eius consequatur qui debitis et perferendis."
      }
    ]
  };

  it("renders AddNote form without issues", () => {
    const select = shallow(
      <Context.Provider value={contextType}>
        <AddNote />
      </Context.Provider>
    );
    expect(select).toMatchSnapshot();
  });
});
