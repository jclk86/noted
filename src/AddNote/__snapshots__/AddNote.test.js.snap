// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddNote Component renders AddNote form without issues 1`] = `ShallowWrapper {}`;
